src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -520,263 +523,198 @@ private unsafe string[] GetValueNamesCore(int values)
+                            };
+
+                        case Interop.Advapi32.RegistryValues.REG_SZ:
+                        case Interop.Advapi32.RegistryValues.REG_EXPAND_SZ:
+                        case Interop.Advapi32.RegistryValues.REG_MULTI_SZ:
                             {
-                                throw new IOException(SR.Arg_RegGetOverflowBug, e);
-                            }
-                        }
-                        char[] blob = new char[datasize / 2];
-
-                        Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref datasize);
-
-                        if (blob.Length > 0 && blob[blob.Length - 1] == (char)0)
-                        {
-                            data = new string(blob, 0, blob.Length - 1);
-                        }
-                        else
-                        {
-                            // in the very unlikely case the data is missing null termination,
-                            // pass in the whole char[] to prevent truncating a character
-                            data = new string(blob);
-                        }
+                                // Handle the case where the registry contains an odd-byte length (corrupt data?)
+                                // by increasing the data by a single zero byte.
+                                if (dataLength % 2 == 1)
+                                {
+                                    if (dataLength == int.MaxValue)
+                                    {
+                                        throw new IOException(SR.Arg_RegValueTooLarge);
+                                    }
 
-                        if (!doNotExpand)
-                        {
-                            data = Environment.ExpandEnvironmentVariables((string)data);
-                        }
-                    }
-                    break;
-                case Interop.Advapi32.RegistryValues.REG_MULTI_SZ:
-                    {
-                        if (datasize % 2 == 1)
-                        {
-                            // handle the case where the registry contains an odd-byte length (corrupt data?)
-                            try
-                            {
-                                datasize = checked(datasize + 1);
-                            }
-                            catch (OverflowException e)
-                            {
-                                throw new IOException(SR.Arg_RegGetOverflowBug, e);
-                            }
-                        }
-                        char[] blob = new char[datasize / 2];
+                                    if (dataLength >= span.Length)
+                                    {
+                                        byte[] newPooled = ArrayPool<byte>.Shared.Rent(dataLength + 1);
```

Comment:
is the rented array zeroed? as you're not necessarily copying over all of it.

Reply:
It's not zero'd. Where am I reading from an uninitialized portion of the array?

Reply:
I thought you were not writing to all of it. If your not reading the remainder then fine. I can't look at the code right now as I'm on my phone.

Reply:
It may not write to all of it, but it should only ever read from what's written.